src/routing/kad-dht.md

```diff
@@ -0,0 +1,666 @@
+    int32 clusterLevelRaw = 10; // NOT USED
+
+    // Used to specify the key associated with this message.
+    // PUT_VALUE, GET_VALUE, ADD_PROVIDER, GET_PROVIDERS
+    bytes key = 2;
+
+    // Used to return a value
+    // PUT_VALUE, GET_VALUE
+    Record record = 3;
+
+    // Used to return peers closer to a key in a query
+    // GET_VALUE, GET_PROVIDERS, FIND_NODE
+    repeated Peer closerPeers = 8;
+
+    // Used to return Providers
+    // GET_VALUE, ADD_PROVIDER, GET_PROVIDERS
+    repeated Peer providerPeers = 9;
+}
+```
+
+These are the requirements for each `MessageType`:
+
+* `FIND_NODE`: In the request `key` must be set to the binary `PeerId` of the
+node to be found. In the response `closerPeers` is set to the DHT Server's `k`
+closest `Peer`s.
+
+* `GET_VALUE`: In the request `key` is an unstructured array of bytes.
+`closerPeers` is set to the `k` closest peers. If `key` is found in the
+datastore `record` is set to the value for the given key.
+
+* `PUT_VALUE`: In the request `record` is set to the record to be stored and
+`key` on `Message` is set to equal `key` of the `Record`. The target node
+validates `record`, and if it is valid, it stores it in the datastore and as a
+response echoes the request.
+
+* `GET_PROVIDERS`: In the request `key` is set to the multihash contained in
+the target CID. The target node returns the known `providerPeers` (if any) and
+the `k` closest known `closerPeers`.
+
+* `ADD_PROVIDER`: In the request `key` is set to the multihash contained in the
+target CID. The target node verifies `key` is a valid multihash, all
+`providerPeers` matching the RPC sender's PeerID are recorded as providers.
+
+* `PING`: Deprecated message type replaced by the dedicated [ping
+protocol](https://github.com/libp2p/specs/blob/master/ping/ping.md).
+
+If a DHT server receives an invalid request, it simply closes the libp2p stream
+without responding.
+
+## Backpressure
+
+TBD
+
+## Client Optimizations
+
+### LAN DHT Swarms
+
+Fine to store private multiaddresses in the routing table and serve them to
+other nodes in the same LAN DHT swarm.
+
+### Checking peer behaviour before adding to routing table
+
+Make a `FIND_NODE` request and inspect response before adding node to RT.
+Followed https://blog.ipfs.tech/2023-ipfs-unresponsive-nodes/
+
+## libp2p Kademlia DHT Implementations
+
+* Go: [`libp2p/go-libp2p-kad-dht`](https://github.com/libp2p/go-libp2p-kad-dht)
+* JS:
+[libp2p/kad-dht](https://github.com/libp2p/js-libp2p/tree/main/packages/kad-dht)
+* Rust:
+[libp2p-kad](https://github.com/libp2p/rust-libp2p/tree/master/protocols/kad)
+
+---
+
+## References
+
+[0]: Maymounkov, P., & Mazières, D. (2002). Kademlia: A Peer-to-Peer
+Information System Based on the XOR Metric. In P. Druschel, F. Kaashoek, & A.
+Rowstron (Eds.), Peer-to-Peer Systems (pp. 53–65). Berlin, Heidelberg: Springer
+Berlin Heidelberg. [DOI](https://doi.org/10.1007/3-540-45748-8_5)
+[pdf](https://www.scs.stanford.edu/~dm/home/papers/kpos.pdf)
```

Comment:
not sure what is the best format for to cite papers with the `spec-generator`

Reply:
yeah, sadly `xref` system only works for web specs.  I've renamed section to "Bibliography" for now, and made manual `[0]` links – we will make this better at some point, but dont want t block this PR on it.